### ISSUE TITLE
Item transformation bug

### ISSUE THREAD

**Issue by limwa** [AllTheMods/SLOP2-slice-of-pie-2#31]
Sorry for the title is so generic, but I really don't know how to describe this bug any better.
It has happened twice in the span of a week on a server I play with my friends.

## The bug

We have our inferium essence stored on a black hole unit inside a black hole controller, that is in turn connected to our AE2 system using a storage bus.
On the first time the bug occurred, we were using AE2 to craft a few supremium watering cans. We had 300K inferium essence. Mid-crafting, we cancelled the crafting. When we got to the ME Terminal, we saw we had 300K tertium essence.
The second time the bug occurred was today. We were using AE2 to craft an Oblivion Shard. We had 5.1M inferium essence. Mid-crafting, we cancelled the crafting. When we got to the ME Terminal, we saw we had 5.1M Starcaller swords.

I'm asking for help here because I don't know of any way to troubleshoot this issue (since this bug could be caused by AE2, ~~MysticalAgriculture~~ or even Industrial Foregoing) and because it could potentially be game-breaking for other players playing the same modpack.

**Comment by limwa** [comment]
The bug just happened again. We were crafting a 16M storage cell and all of our iron (it was stored on a Black Hole Unit) turned into calculation processor when we cancelled the crafting.

**Comment by limwa** [comment]
With this, we can rule out Mystical Agriculture as a possible cause.

**Comment by MadWomen** [comment]
very interesting bug indeed, i will be looking into it.

**Comment by limwa** [comment]
So, I've managed to replicate the bug in single-player mode as well. 

## The Setup

![2021-09-19_19 43 49](https://user-images.githubusercontent.com/13498603/133939110-b14f0625-18d3-4e1f-b257-318bdc2494f8.png)

The setup is almost the same as what we have on the server.
The command blocks are there to fill the chest, every tick, with 1 stack of redstone, iron, osmium, inferium, and diamonds (you could also use drawers with creative vending upgrades).
The ME Drive has a 64k Storage Cell and (I don't know if this matters) everything from Applied Energistics has been upgraded with 4 Accelerator Cards.

Using the Crafting Terminal, I've also stored 5 Master Infusion Crystals on the Storage Cell (important later).

The storage bus behind the black hole controller has the default configuration, except it has a priority of 2000.

![2021-09-19_19 48 20](https://user-images.githubusercontent.com/13498603/133939234-957d9c62-ed5c-481e-b01b-fdbd2a4a291b.png)

The Black Hole Controller has 5 Supremium Black Hole Units inside: one with redstone, another with iron, another with osmium, another with inferium, and another with diamonds.

Finally, I've added 5 patterns to the network:

![2021-09-19_19 54 09](https://user-images.githubusercontent.com/13498603/133939405-3520d283-8553-4bd7-b546-311c038799d0.png)

The reason I've used these recipes is 1. they've bugged at least once in the past and 2. they take long to care and so I have more time to cancel the recipe since cancelling the crafting is an important step.

## Making the bug happen

1. Craft a good quantity of Insanium Essence (32 or more), since it takes a long time to craft

![2021-09-19_20 00 22](https://user-images.githubusercontent.com/13498603/133939577-e9d287b5-1daa-4793-b45c-3067ad0e56d6.png)

2. Wait for a few seconds and, then, cancel the crafting

![2021-09-19_20 01 35](https://user-images.githubusercontent.com/13498603/133940236-21017903-ce4a-44c1-be46-57d04bcbd279.png)

3. Go to your terminal. You should have a huge quantity of something else waiting for you. In my case, I got 54K Master Infusion Crystals (all the Inferium in the Black Hole Unit got replaced by Master Infusion Crystals).

![2021-09-19_20 03 25](https://user-images.githubusercontent.com/13498603/133939661-df760d85-9280-4d84-b343-09affcf9ccdd.png)
![2021-09-19_20 03 47](https://user-images.githubusercontent.com/13498603/133939672-37cfb1af-4531-407c-bc4e-c71158023646.png)

This bug is very weird and I'll keep investigating to see if I can figure out what mod is causing it.

**Note:** I am aware that there are items on JEI from Twilight Forest and that Twilight Forest is not in this modpack. This is because I and my friends play on ATM-SLOP2 + CC: Tweaked + Twilight Forest. Anyhow, **I have tried this same exact procedure on the latest version of ATM-SLOP2 and the bug still happens**.

**Comment by limwa** [comment]
**Update:** The bug doesn't seem to happen when Storage Drawers are used.
**Update 2:** The bug also doesn't seem to happen when only Black Hole Units are used (without the Black Hole Controller).

Looks like there is a bug with the Black Hole Controller.

**Comment by lLuckyLukel** [comment]
@limwa can you please provide your world save download for further testing?

**Comment by limwa** [comment]
@lLuckyLukel Here it goes!
[Bug Demo.zip](https://github.com/AllTheMods/SLOP2-slice-of-pie-2/files/7230284/Bug.Demo.zip)

**Comment by lLuckyLukel** [comment]
@limwa Oh yes, sir, thank you very much! I really appreciate that you made it as a tutorial, I was actually really amazed seeing this first time loading that save. big surprise XD

**Comment by limwa** [comment]
Hahaha, you're welcome. Thought it would be easier for me and for other people to verify the bug.

**Comment by lLuckyLukel** [comment]
So yeah, I'm not sure exactly what's causing the issue, still playing around with it on the server (it works), but actually there was kinda similar duplication bug, just using only one Black Hole Unit, where you can overwrite, let's say 10k cobblestone into 10k diamond with only 2-3 clicks. But that one was already fixed on the server, so maybe it's a kind of IF + AE2 bypussing our fix

**Comment by limwa** [comment]
Yeah, I also don't know exactly what's causing this. 
I've noticed a few things during my testing, though:
1. It only happened when I had 3 co-processors (if I had more, it would create 1 inferium essence at a time instead of creating all of the prudentium essence needed first, then all of the tertium essence needed, ...)
2. The process (crafting ~~or other machines~~) has to be really fast. If it is slower, although still possible, the bug becomes much harder to pull off. Maybe it has something to do with interrupting the request while the crafted items are leaving the Molecular Assemblers and entering the ME Network.
3. It only seems to work with Black Hole Controllers.

**EDIT**: I crossed "or other machines" because I'm not sure if I'm remembering things correctly.

**Comment by lLuckyLukel** [comment]
And one more thing, I also noticed one interesting moment, this happens if the time interval between the "Start" of the crafting process and the "Close" is ~10 seconds, I tried to Close faster, but seems like it only works within 9-10 sec

**Comment by limwa** [comment]
Yes, however, I've also noticed that if you request more items (like 32 insanium essence), it gives you a bigger window to cancel the request
